src/Application/Actions/Donations/Update.php

```diff
@@ -113,6 +113,20 @@ protected function action(Request $request, Response $response, array $args): Re
                     );
                 }
 
+                if ($donationData->autoConfirmFromCashBalance && $donation->getPaymentMethodType() !== 'customer_balance') {
```

Comment:
not necessarily now, but at some point it might be good to make payment method type an enum, with just the two cases `card` and `customerBalance` - and we might be able to share the enum between the domain model and the DTO. Or we could more cheaply leave it as-is at runtime but type it statically as `'card'|'customer_balance'` instead of `string`. Could also make it an enum in the db scehma.

Reply:
Why would a non customer balance donation have `autoConfirmFromCashBalance` set? Would it have to be a bug in the client? I see our frontend only sets this in [finaliseCashBalancePurchase](https://github.com/thebiggive/donate-frontend/blob/853dd7101148255a57e0118a3e36e682f277994b/src/app/donation.service.ts#L135)

Reply:
@bdsl I'm not totally sure. I think probably the best guess I have right now is the 3rd of the scenarios I suggested testing:

> 1. donating logged in in 2 different browsers on a Customer with credit: setting up a donation in each (using the same balance), completing 1 (which should work) and then attempting to complete the 2nd (which should give an error)
> 2. starting a donation logged out, then [if possible] returning to the start to log into an account with a balance before completing the original donation
> 3. starting a donation logged in but with £0 balance, then funding the account in Stripe test mode, refreshing and then choosing to continue the previously set up donation [which I think may also error at the end]

Reply:
> not necessarily now, but at some point it might be good to make payment method type an enum, with just the two cases `card` and `customerBalance` - and we might be able to share the enum between the domain model and the DTO. Or we could more cheaply leave it as-is at runtime but type it statically as `'card'|'customer_balance'` instead of `string`. Could also make it an enum in the db scehma.

I'm nervous about the testing and work to move an already-live field to an enum at DB level, unless there is a substantial benefit relevant to this change.

I've started using a PHP `enum` for the property, along with some possible or necessary related simplifications, on https://github.com/thebiggive/matchbot/pull/550/commits/2670d7d922754e93678e2ceb7b1a67e8a7193127

I had to `@psalm-suppress MissingConstructor` on the HTTP model in order to use an enum in an object without a constructor. Not sure if there's a more elegant way to handle that?

Reply:
Will have to continue this tomorrow - the missing constructor complaint isn't because https://github.com/thebiggive/matchbot/commit/2670d7d922754e93678e2ceb7b1a67e8a7193127 changes the string to an enum, it's because it removes the default value from the declaration. Declaring the property as `public PaymentMethodType $paymentMethodType == PaymentMethodType::card`; should make that go away.

Whether or not it should have that default is a separate question to whether it should be an enum instead of a string.

Btw I think it will need to be a string-backed enum to let doctrine persist it to the db.

Reply:
@bdsl I think I realised last week that there were already errors in the baseline for other properties on the same model. So effectively we were suppressing it on each other property already.